File: backend/src/voice/voice.service.ts
import { Injectable, Logger, UnauthorizedException, NotFoundException } from '@nestjs/common';
import { SupabaseService } from '../supabase/supabase.module';

@Injectable()
export class VoiceService {
  private readonly logger = new Logger(VoiceService.name);

  constructor(private readonly supabaseService: SupabaseService) {}

  async validateVoiceRoomAccess(roomId: string, userId: string): Promise<boolean> {
    try {
      const { data: group, error } = await this.supabaseService
        .getClient()
        .from('groups')
        .select(`
          id,
          channel_id,
          group_type,
          access_type,
          group_members!inner(user_id)
        `)
        .eq('id', roomId)
        .eq('group_type', 'voice')
        .eq('group_members.user_id', userId)
        .single();

      if (error || !group) {
        this.logger.warn(`User ${userId} denied access to voice room ${roomId}`);
        return false;
      }

      return true;
    } catch (error) {
      this.logger.error(`Error validating voice room access: ${error.message}`);
      return false;
    }
  }

  async getVoiceRoomInfo(roomId: string, userId: string) {
    const hasAccess = await this.validateVoiceRoomAccess(roomId, userId);
    if (!hasAccess) {
      throw new UnauthorizedException('Access denied to voice room');
    }

    const { data: group, error } = await this.supabaseService
      .getClient()
      .from('groups')
      .select(`
        id,
        name,
        description,
        channel_id,
        access_type,
        created_at,
        group_members(
          user_id,
          role,
          joined_at,
          users!group_members_user_id_fkey(
            id,
            username,
            avatar_url,
            full_name
          )
        )
      `)
      .eq('id', roomId)
      .single();

    if (error || !group) {
      throw new NotFoundException('Voice room not found');
    }

    return group;
  }

  async getUserVoiceGroups(userId: string) {
    const { data, error } = await this.supabaseService
      .getClient()
      .from('groups')
      .select(`
        id,
        name,
        description,
        channel_id,
        access_type,
        created_at,
        group_members!inner(
          user_id,
          role,
          joined_at
        )
      `)
      .eq('group_type', 'voice')
      .eq('group_members.user_id', userId)
      .order('created_at', { ascending: true });

    if (error) {
      throw new Error(`Failed to fetch voice groups: ${error.message}`);
    }

    return data;
  }

  async createVoiceRoom(channelId: string, creatorUserId: string, createData: {
    name: string;
    description?: string;
    access_type?: 'PUBLIC' | 'PRIVATE';
  }) {
    // Verify user is channel member
    const { data: member } = await this.supabaseService
      .getClient()
      .from('channel_members')
      .select('role')
      .eq('channel_id', channelId)
      .eq('user_id', creatorUserId)
      .single();

    if (!member || (member.role !== 'OWNER' && member.role !== 'ADMIN')) {
      throw new UnauthorizedException('Only channel admins can create voice rooms');
    }

    const { data, error } = await this.supabaseService
      .getClient()
      .from('groups')
      .insert({
        channel_id: channelId,
        name: createData.name,
        description: createData.description,
        group_type: 'voice',
        access_type: createData.access_type || 'PUBLIC',
      })
      .select()
      .single();

    if (error) {
      throw new Error(`Failed to create voice room: ${error.message}`);
    }

    // Add creator as admin
    await this.supabaseService
      .getClient()
      .from('group_members')
      .insert({
        group_id: data.id,
        user_id: creatorUserId,
        role: 'ADMIN',
      });

    this.logger.log(`Voice room ${data.id} created by user ${creatorUserId}`);
    return data;
  }
}

File: goback/services/voice_service.go
package services

import (
	"database/sql"
	"fmt"

	"github.com/Hamza-agha-tec/goback/models"
)

type VoiceService struct {
	db       *sql.DB
	supabase *SupabaseClient
}

func NewVoiceService(db *sql.DB, supabaseURL, supabaseKey string) *VoiceService {
	return &VoiceService{
		db:       db,
		supabase: NewSupabaseClient(supabaseURL, supabaseKey),
	}
}

func (s *VoiceService) GetUserVoiceGroups(userID string) ([]*models.VoiceRoom, error) {
	// First get room IDs where user is a member
	var members []models.VoiceRoomMember
	err := s.supabase.GetClient().DB.From("voice_room_members").
		Select("room_id").
		Eq("user_id", userID).
		Execute(&members)

	if err != nil {
		return nil, fmt.Errorf("failed to query user room memberships: %w", err)
	}

	if len(members) == 0 {
		return []*models.VoiceRoom{}, nil
	}

	// Extract room IDs
	roomIDs := make([]string, len(members))
	for i, member := range members {
		roomIDs[i] = member.RoomID
	}

	// Get voice rooms
	var rooms []models.VoiceRoom
	err = s.supabase.GetClient().DB.From("voice_rooms").
		Select("*").
		In("id", roomIDs).
		Execute(&rooms)

	if err != nil {
		return nil, fmt.Errorf("failed to query voice rooms: %w", err)
	}

	// Convert to pointer slice
	var roomPtrs []*models.VoiceRoom
	for i := range rooms {
		roomPtrs = append(roomPtrs, &rooms[i])
	}

	return roomPtrs, nil
}

func (s *VoiceService) GetVoiceRoomInfo(roomID, userID string) (*models.VoiceRoom, error) {
	// Check if user has access to the room
	hasAccess, err := s.validateVoiceRoomAccess(roomID, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to validate room access: %w", err)
	}
	if !hasAccess {
		return nil, fmt.Errorf("access denied")
	}

	var rooms []models.VoiceRoom
	err = s.supabase.GetClient().DB.From("voice_rooms").
		Select("*").
		Eq("id", roomID).
		Execute(&rooms)

	if err != nil {
		return nil, fmt.Errorf("failed to query voice room: %w", err)
	}

	if len(rooms) == 0 {
		return nil, fmt.Errorf("voice room not found")
	}

	return &rooms[0], nil
}

func (s *VoiceService) CreateVoiceRoom(channelID, userID string, createData map[string]interface{}) (*models.VoiceRoom, error) {
	name := createData["name"].(string)
	description, _ := createData["description"].(string)
	accessType := "PUBLIC"
	if at, ok := createData["access_type"].(string); ok {
		accessType = at
	}

	roomData := map[string]interface{}{
		"channel_id":  channelID,
		"name":        name,
		"description": description,
		"access_type": accessType,
		"owner_id":    userID,
		"is_active":   true,
	}

	var rooms []models.VoiceRoom
	err := s.supabase.GetClient().DB.From("voice_rooms").
		Insert(roomData).
		Execute(&rooms)

	if err != nil {
		return nil, fmt.Errorf("failed to create voice room: %w", err)
	}

	if len(rooms) == 0 {
		return nil, fmt.Errorf("failed to create voice room: no room returned")
	}

	room := &rooms[0]

	// Add owner as member
	memberData := map[string]interface{}{
		"room_id":     room.ID,
		"user_id":     userID,
		"is_muted":    false,
		"is_deafened": false,
		"is_speaking": false,
	}

	var members []models.VoiceRoomMember
	err = s.supabase.GetClient().DB.From("voice_room_members").
		Insert(memberData).
		Execute(&members)

	if err != nil {
		return nil, fmt.Errorf("failed to add owner as room member: %w", err)
	}

	return room, nil
}

func (s *VoiceService) ValidateVoiceRoomAccess(roomID, userID string) (bool, error) {
	return s.validateVoiceRoomAccess(roomID, userID)
}

func (s *VoiceService) validateVoiceRoomAccess(roomID, userID string) (bool, error) {
	// Check if user is a member of the voice room
	var members []models.VoiceRoomMember
	err := s.supabase.GetClient().DB.From("voice_room_members").
		Select("id").
		Eq("room_id", roomID).
		Eq("user_id", userID).
		Execute(&members)

	if err != nil {
		return false, fmt.Errorf("failed to check room membership: %w", err)
	}

	if len(members) > 0 {
		return true, nil
	}

	// Check if user is a member of the parent channel and room is public
	var rooms []models.VoiceRoom
	err = s.supabase.GetClient().DB.From("voice_rooms").
		Select("id").
		Eq("id", roomID).
		Eq("access_type", "PUBLIC").
		Execute(&rooms)

	if err != nil {
		return false, fmt.Errorf("failed to check room access type: %w", err)
	}

	if len(rooms) == 0 {
		return false, nil
	}

	// Check channel membership (complex JOIN, using raw SQL for now)
	var isChannelMember bool
	err = s.db.QueryRow(`
		SELECT EXISTS(
			SELECT 1 FROM channel_members 
			WHERE channel_id = (SELECT channel_id FROM voice_rooms WHERE id = $1) AND user_id = $2
		)
	`, roomID, userID).Scan(&isChannelMember)

	if err != nil {
		return false, fmt.Errorf("failed to check channel membership: %w", err)
	}

	return isChannelMember, nil
}
